File: android/app/src/main/java/com/gestorfinances/app/domain/rules/GoalProgress.kt
package com.gestorfinances.app.domain.rules

import java.time.LocalDate

/** How a savings goal knows how much it has saved. */
enum class GoalFundingMode(val dbValue: String) {
    /** Progress follows the canonical value of one account reserved for this goal alone. */
    DEDICATED_ACCOUNT("dedicated_account"),

    /** Progress is the signed sum of dated planning allocations, so goals can share an account. */
    ALLOCATIONS("allocations"),
    ;

    companion object {
        fun fromDb(value: String): GoalFundingMode =
            entries.firstOrNull { it.dbValue == value } ?: error("Unknown goal funding mode: $value")
    }
}

/**
 * Progress and required monthly pace for one savings goal, locked by
 * `shared/golden/goal_progress.json`.
 *
 * Reading a goal never moves the ledger: an allocation reserves meaning only, so nothing here
 * touches account flow, actual income or expense, debt, or net worth. [savedCents] must equal
 * `v_goal_progress.saved_cents` for the same goal; this rule adds only the pace, which is
 * awkward to express in SQL.
 */
data class GoalProgress(
    val savedCents: Long,
    val remainingCents: Long,
    val reached: Boolean,
    val monthsRemaining: Int?,
    val monthlyPaceCents: Long?,
    val overdue: Boolean,
) {
    companion object {
        fun evaluate(
            fundingMode: GoalFundingMode,
            targetAmountCents: Long,
            accountBalanceCents: Long,
            allocationCents: List<Long>,
            targetDate: LocalDate?,
            today: LocalDate,
        ): GoalProgress {
            val saved = when (fundingMode) {
                GoalFundingMode.DEDICATED_ACCOUNT -> accountBalanceCents
                GoalFundingMode.ALLOCATIONS -> allocationCents.sum()
            }
            return of(saved, targetAmountCents, targetDate, today)
        }

        /** Pace and remainder for a goal whose saved amount already comes from `v_goal_progress`. */
        fun of(
            savedCents: Long,
            targetAmountCents: Long,
            targetDate: LocalDate?,
            today: LocalDate,
        ): GoalProgress {
            val remaining = (targetAmountCents - savedCents).coerceAtLeast(0)
            val reached = remaining == 0L

            if (targetDate == null) {
                return GoalProgress(
                    savedCents = savedCents,
                    remainingCents = remaining,
                    reached = reached,
                    monthsRemaining = null,
                    monthlyPaceCents = null,
                    overdue = false,
                )
            }

            // Calendar months counted inclusively from this month through the target's month, so a
            // target inside the current month still leaves one month to fund it.
            val months = (monthsBetween(today, targetDate) + 1).coerceAtLeast(0)
            if (reached) {
                return GoalProgress(
                    savedCents = savedCents,
                    remainingCents = 0,
                    reached = true,
                    monthsRemaining = months,
                    monthlyPaceCents = 0,
                    overdue = false,
                )
            }

            // Rounding up keeps the target reachable: exact division would leave a cent short.
            val pace = if (months == 0) remaining else ceilingDivide(remaining, months)
            return GoalProgress(
                savedCents = savedCents,
                remainingCents = remaining,
                reached = false,
                monthsRemaining = months,
                monthlyPaceCents = pace,
                overdue = months == 0,
            )
        }

        private fun monthsBetween(from: LocalDate, to: LocalDate): Int =
            (to.year * 12 + to.monthValue) - (from.year * 12 + from.monthValue)

        private fun ceilingDivide(value: Long, divisor: Int): Long =
            (value + divisor - 1) / divisor
    }
}

File: windows/GestorFinances.Tests/Domain/Rules/GoalProgress.cs
namespace GestorFinances.Domain.Rules;

public enum GoalFundingMode
{
    DedicatedAccount,
    Allocations
}

public sealed record GoalProgressInput(
    GoalFundingMode FundingMode,
    long TargetAmountCents,
    long AccountBalanceCents,
    IReadOnlyList<long> AllocationCents,
    DateOnly? TargetDate,
    DateOnly Today);

public sealed record GoalProgressResult(
    long SavedCents,
    long RemainingCents,
    bool Reached,
    int? MonthsRemaining,
    long? MonthlyPaceCents,
    bool Overdue);

/// <summary>
/// Savings-goal progress and required monthly pace. Locked by shared/golden/goal_progress.json.
/// Reading a goal never moves the ledger: allocations reserve meaning only.
/// </summary>
public static class GoalProgress
{
    public static GoalProgressResult Evaluate(GoalProgressInput input)
    {
        var saved = input.FundingMode switch
        {
            GoalFundingMode.DedicatedAccount => input.AccountBalanceCents,
            GoalFundingMode.Allocations => input.AllocationCents.Sum(),
            _ => throw new ArgumentOutOfRangeException(nameof(input), "Unknown goal funding mode")
        };

        var remaining = Math.Max(input.TargetAmountCents - saved, 0);
        var reached = remaining == 0;

        if (input.TargetDate is not { } targetDate)
        {
            return new GoalProgressResult(saved, remaining, reached, null, null, Overdue: false);
        }

        // Calendar months are counted inclusively from today's month through the target's month,
        // so a target today or later this month leaves one period; past dates leave none.
        var monthsRemaining = targetDate < input.Today ? 0 : MonthsBetween(input.Today, targetDate) + 1;
        if (monthsRemaining < 0)
        {
            monthsRemaining = 0;
        }

        if (reached)
        {
            return new GoalProgressResult(saved, remaining, Reached: true, monthsRemaining, 0, Overdue: false);
        }

        // Rounding up keeps the target reachable: an exact division would leave a cent short.
        var pace = monthsRemaining == 0 ? remaining : CeilingDivide(remaining, monthsRemaining);
        return new GoalProgressResult(saved, remaining, Reached: false, monthsRemaining, pace, monthsRemaining == 0);
    }

    private static int MonthsBetween(DateOnly from, DateOnly to) =>
        ((to.Year * 12) + to.Month) - ((from.Year * 12) + from.Month);

    private static long CeilingDivide(long value, int divisor) =>
        (value + divisor - 1) / divisor;
}
